File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.py
# TC- O(logn)
# SC- O(1)

class Solution:
    def findMin(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        
        left = 0
        right = len(nums)-1

        # if not rotated
        if nums[left]<nums[right]:
            return nums[0]
        
        # if rotated
        while(left<right):
            mid = left+ (right-left)//2
            if nums[mid]>nums[mid+1]: # 3 4 5 1 2
                return nums[mid+1]
            
            elif nums[mid]<nums[mid-1]: # 4 5 1 2 3
                return nums[mid]

            elif nums[left]>nums[mid]: # 5 1 2 3 4
                right = mid
            
            else:   # 2 3 4 5 1
                left = mid+1

File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.cpp
// TC- O(logn)
// sc- o(1)

class Solution {
public:
    int findMin(vector<int>& nums) {
        // edge case
        if(nums.size()==1)
            return nums[0];


        int lo = 0;
        int hi = nums.size()-1;

        // array is not rotated
        if(nums[hi]>nums[lo])
            return nums[lo];
        
        // array is rotated
        while(lo<hi){
            int mid = lo +(hi-lo)/2;
            if(nums[mid]>nums[hi])
                lo=mid+1;
            else
                hi=mid;
        }
        return nums[lo];
    }
};
// 1 2 3 4 5
// 3 4 5 1 2
// 1 2 3 4 5

